core/contextmenu_registry.js

```diff
@@ -0,0 +1,138 @@
+ */
+'use strict';
+
+/**
+ * @name Blockly.ContextMenuRegistry
+ * @namespace
+ */
+goog.provide('Blockly.ContextMenuRegistry');
+
+/**
+ * Where this menu item should be rendered. If the menu item should be rendered in multiple
+ * scopes, e.g. on both a block and a workspace, it should be registered for each scope.
+ * @enum {string}
+ */
+Blockly.ContextMenuRegistry.ScopeType = {
+  BLOCK: 'block',
+  WORKSPACE: 'workspace',
+};
+
+/**
+ * The actual workspace/block where the menu is being rendered. This is passed to callback and
+ * displayText functions that depend on this information.
+ * @typedef {{
+ *    block: (Blockly.BlockSvg|undefined),
+ *    workspace: (Blockly.WorkspaceSvg|undefined),
+ * }}
+ */
+Blockly.ContextMenuRegistry.Scope;
+
+/**
+ * A menu item as entered in the registry.
+ * @typedef {{
+ *    callback: function(*),
+ *    scopeType: !Blockly.ContextMenuRegistry.ScopeType,
+ *    displayText: ((function(!Blockly.ContextMenuRegistry.Scope):string)|string),
+ *    preconditionFn: function(!Blockly.ContextMenuRegistry.Scope):string,
+ *    weight: number,
+ *    id: string,
+ * }}
+*/
+Blockly.ContextMenuRegistry.RegistryItem;
+
+/**
+ * A menu item as presented to contextmenu.js.
+ * @typedef {{
+ *    text: string,
+ *    enabled: boolean,
+ *    callback: function(*),
+ *    scope: !Blockly.ContextMenuRegistry.Scope,
+ * }}
+ */
+Blockly.ContextMenuRegistry.ContextMenuOption;
+
+/**
+ * Registry of all registered RegistryItems, keyed by id.
+ * @type {!Object<string, Blockly.ContextMenuRegistry.RegistryItem>}
+ */
+Blockly.ContextMenuRegistry.registry_ = {};
+
+/**
+ * Registers a RegistryItem.
+ * @param {Blockly.ContextMenuRegistry.RegistryItem} item Context menu item to register.
+ * @throws {Error} if an item with the given id already exists.
+ */
+Blockly.ContextMenuRegistry.register = function(item) {
+  if (Blockly.ContextMenuRegistry.registry_[item.id]) {
+    throw Error('Menu item with id "' + item.id + '" is already registered.');
+  }
+  Blockly.ContextMenuRegistry.registry_[item.id] = item;
+};
+
+/**
+ * Unregisters a RegistryItem with the given id.
+ * @param {string} id The id of the RegistryItem to remove.
+ * @throws {Error} if an item with the given id does not exist.
+ */
+Blockly.ContextMenuRegistry.unregister = function(id) {
+  if (Blockly.ContextMenuRegistry.registry_[id]) {
+    delete Blockly.ContextMenuRegistry.registry_[id];
+  } else {
+    throw new Error('Menu item with id "' + id + '" not found.');
+  }
+};
+
+/**
+ * @param {string} id The id of the RegistryItem to get.
+ * @returns {Blockly.ContextMenuRegistry.RegistryItem} RegistryItem
+ * @throws {Error} if an item with the given id does not exist.
+ */
+Blockly.ContextMenuRegistry.getItem = function(id) {
+  if (Blockly.ContextMenuRegistry.registry_[id]) {
+    return Blockly.ContextMenuRegistry.registry_[id];
+  } else {
+    throw new Error('Menu item with id "' + id + '" not found.');
```

Comment:
I think this should return null, possibly logging a warning, rather than throwing.

Reply:
Done (decided not to log, they can check for null if they care)